importer/balanced/builder.go

```diff
@@ -27,87 +43,114 @@ import (
 // and depth only increases when the tree is full, that is, when
 // the root node has reached the maximum number of links.
 func Layout(db *h.DagBuilderHelper) (ipld.Node, error) {
-	var offset uint64
-	var root *h.UnixfsNode
-	for level := 0; !db.Done(); level++ {
-
-		nroot := db.NewUnixfsNode()
-		db.SetPosInfo(nroot, 0)
-
-		// add our old root as a child of the new root.
-		if root != nil { // nil if it's the first node.
-			if err := nroot.AddChild(root, db); err != nil {
-				return nil, err
-			}
-		}
-
-		// fill it up.
-		if err := fillNodeRec(db, nroot, level, offset); err != nil {
-			return nil, err
-		}
-
-		offset = nroot.FileSize()
-		root = nroot
-
-	}
-	if root == nil {
-		// this should only happen with an empty node, so return a leaf
-		var err error
-		root, err = db.NewLeaf(nil)
+	if db.Done() {
+		// No data, return just an empty node.
+		root, err := db.NewLeafNode(nil)
 		if err != nil {
 			return nil, err
 		}
+		// This works without Filestore support (`ProcessFileStore`).
+		// TODO: Why? Is there a test case missing?
+
+		return db.AddNodeAndClose(root)
 	}
 
-	out, err := db.Add(root)
+	// The first `root` will be a single leaf node with data
+	// (corner case), after that subsequent `root` nodes will
+	// always be internal nodes (with a depth > 0) that can
+	// be handled by the loop.
+	root, fileSize, err := db.NewLeafDataNode()
 	if err != nil {
 		return nil, err
 	}
 
-	err = db.Close()
-	if err != nil {
-		return nil, err
+	// Each time a DAG of a certain `depth` is filled (because it
+	// has reached its maximum capacity of `db.Maxlinks()` per node)
+	// extend it by making it a sub-DAG of a bigger DAG with `depth+1`.
+	for depth := 1; !db.Done(); depth++ {
+
+		// Add the old `root` as a child of the `newRoot`.
+		newRoot := db.NewFSNodeOverDag(ft.TFile)
+		newRoot.AddChild(root, fileSize, db)
+
+		// Fill the `newRoot` (that has the old `root` already as child)
+		// and make it the current `root` for the next iteration (when
+		// it will become "old").
+		root, fileSize, err = fillNodeRec(db, newRoot, depth)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return out, nil
+	return db.AddNodeAndClose(root)
 }
 
-// fillNodeRec will fill the given node with data from the dagBuilders input
-// source down to an indirection depth as specified by 'depth'
-// it returns the total dataSize of the node, and a potential error
+// fillNodeRec will fill the given internal (non-leaf) `node` with either
+// more internal nodes or leaf data nodes (when the final `depth` of 1 is
+// reached) containing the file data passed from the `DagBuilderHelper`.
+//
+// If `node` is `nil` it means we're descending into new territory so
+// just create a new node to be filled (recursion call); if it's not
+// `nil` it has been called from `Layout` where the DAG has been extended
+// to a new root, that already has a child (the old root), to be filled
+// with more children (up to `db.Maxlinks()`).
```

Comment:
A diagram might help again here:

```
// `node` represents a new root that is being expanded like "Root 2" in the
// diagram below. If it is `nil`, then this function should create a new node
// like "Root 3" in the diagram below. (It should only be `nil` when calling
// itself recursively, having started with a node like "Root 2".)
// 
//                         +-------------+
//                         |   Root 2    |
//                         +-------------+
//                               |       
//                 +-------------+ - - - - - - +
//                 |                           |
//          +-------------+             + - - - - - - +
//          |   Root 1    |             |   Root 3    |
//          +-------------+             + - - - - - - +
//                 |                           |       
//          +------+------+             + - - -+ - - -+
//          |             |             |             |
//     +=========+   +=========+   + - - - - +   + - - - - +
//     | Chunk 1 |   | Chunk 2 |   | Chunk 3 |   | Chunk 4 |
//     +=========+   +=========+   + - - - - +   + - - - - +
//
// The `Layout` function is responsible for creating new roots like "Root 2"
// when more data doesn't fit in the current root. It call this function, which
// is responsible for creating the new branch like "Root 3" by calling itself
// recursively.
```

Reply:
Yes, the package doc should have a filled balanced DAG to make sure we leave a clear idea of what *balanced* implies, and this function can have an incomplete DAG in the process of being filled.

> The `Layout` function is responsible for creating new roots

Yes, that is a very important conceptual distinction, `Layout` is the only function that creates root nodes (and expands the DAG skeleton, while the `fill` call fleshes it out).